File: src/bindings/ruby/examples/metrics_example.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

#
# NIMCP Metrics Example
# Demonstrates metrics collection for Tableau and PowerBI
#

require_relative '../lib/nimcp'

begin
  # Initialize NIMCP
  NIMCP.init
  puts "NIMCP Version: #{NIMCP.version}"

  # Create metrics collector
  metrics = NIMCP::MetricsCollector.new(
    directory: './ruby_metrics',
    format: :csv
  )
  puts "\nMetrics collector created"

  # Record various metrics
  puts "\nRecording metrics..."

  # Counter metrics
  metrics.record_counter('forward_passes', 1000, category: :performance)
  metrics.record_counter('training_examples', 5000, category: :learning)

  # Gauge metrics
  metrics.record_gauge('learning_rate', 0.001, category: :learning)
  metrics.record_gauge('accuracy', 0.95, category: :inference)
  metrics.record_gauge('memory_usage_mb', 256.5, category: :memory)

  # Timer metrics
  metrics.record_timer('forward_time_ms', 15.5, category: :performance)
  metrics.record_timer('backward_time_ms', 22.3, category: :performance)

  # Timing block example
  metrics.timer('data_processing', category: :performance) do
    # Simulate some work
    sleep(0.1)
    puts "Processing data..."
  end

  # Simulate training loop with metrics
  puts "\nSimulating training loop with metrics..."
  5.times do |epoch|
    metrics.timer("epoch_#{epoch}", category: :learning) do
      # Record epoch-specific metrics
      metrics.record_gauge('current_epoch', epoch, category: :learning)
      metrics.record_gauge('loss', 0.5 / (epoch + 1), category: :learning)

      # Simulate training time
      sleep(0.05)
    end
  end

  # Get and display statistics
  puts "\nMetrics Statistics:"
  stats = metrics.stats
  puts "  Total metrics recorded: #{stats['total_recorded']}"
  puts "  Total metrics flushed: #{stats['total_flushed']}"
  puts "  Buffer count: #{stats['buffer_count']}"

  # Flush metrics to disk
  puts "\nFlushing metrics..."
  count = metrics.flush
  puts "  Flushed #{count} metrics to disk"

  # Export for Tableau
  tableau_file = './ruby_metrics_tableau.csv'
  puts "\nExporting to Tableau CSV: #{tableau_file}"
  metrics.export_tableau_csv(tableau_file)

  # Export for PowerBI
  powerbi_file = './ruby_metrics_powerbi.json'
  puts "Exporting to PowerBI JSON: #{powerbi_file}"
  metrics.export_powerbi_json(powerbi_file)

  # Change directory and record more metrics
  puts "\nChanging metrics directory..."
  metrics.set_directory('./ruby_metrics_alternate')
  metrics.record_counter('additional_operations', 100, category: :system)

  # Get updated statistics
  stats = metrics.stats
  puts "\nFinal Statistics:"
  puts "  Total metrics recorded: #{stats['total_recorded']}"

  # Clean up
  metrics.destroy
  puts "\nMetrics collector destroyed"

  # Shutdown NIMCP
  NIMCP.shutdown
  puts "NIMCP shutdown complete"

rescue NIMCP::Error => e
  puts "Error: #{e.message}"
  exit 1
end

File: src/bindings/nodejs/metrics_example.js
#!/usr/bin/env node
/**
 * NIMCP Metrics Example for Node.js
 * Demonstrates metrics collection for Tableau and PowerBI
 */

const nimcp = require('./index');

async function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
    try {
        console.log('NIMCP Metrics Example\n');

        // Create metrics collector
        const metrics = new nimcp.MetricsCollector({
            directory: './nodejs_metrics'
        });
        console.log('Metrics collector created\n');

        // Record various metrics
        console.log('Recording metrics...');

        // Counter metrics
        metrics.recordCounter('forward_passes', 1000, nimcp.METRIC_CATEGORY_PERFORMANCE);
        metrics.recordCounter('training_examples', 5000, nimcp.METRIC_CATEGORY_LEARNING);

        // Gauge metrics
        metrics.recordGauge('learning_rate', 0.001, nimcp.METRIC_CATEGORY_LEARNING);
        metrics.recordGauge('accuracy', 0.95, nimcp.METRIC_CATEGORY_INFERENCE);
        metrics.recordGauge('memory_usage_mb', 256.5, nimcp.METRIC_CATEGORY_MEMORY);

        // Timer metrics
        metrics.recordTimer('forward_time_ms', 15.5, nimcp.METRIC_CATEGORY_PERFORMANCE);
        metrics.recordTimer('backward_time_ms', 22.3, nimcp.METRIC_CATEGORY_PERFORMANCE);

        // Timing example
        const startTime = Date.now();
        await sleep(100);
        const duration = Date.now() - startTime;
        metrics.recordTimer('data_processing', duration, nimcp.METRIC_CATEGORY_PERFORMANCE);
        console.log('Processed data in ' + duration + 'ms');

        // Simulate training loop with metrics
        console.log('\nSimulating training loop with metrics...');
        for (let epoch = 0; epoch < 5; epoch++) {
            const epochStart = Date.now();

            // Record epoch-specific metrics
            metrics.recordGauge('current_epoch', epoch, nimcp.METRIC_CATEGORY_LEARNING);
            metrics.recordGauge('loss', 0.5 / (epoch + 1), nimcp.METRIC_CATEGORY_LEARNING);

            // Simulate training time
            await sleep(50);

            const epochDuration = Date.now() - epochStart;
            metrics.recordTimer('epoch_' + epoch, epochDuration, nimcp.METRIC_CATEGORY_LEARNING);
        }

        // Get and display statistics
        console.log('\nMetrics Statistics:');
        const stats = JSON.parse(metrics.getStats());
        console.log('  Total metrics recorded:', stats.total_recorded);
        console.log('  Total metrics flushed:', stats.total_flushed);
        console.log('  Buffer count:', stats.buffer_count);

        // Flush metrics to disk
        console.log('\nFlushing metrics...');
        const count = metrics.flush();
        console.log('  Flushed ' + count + ' metrics to disk');

        // Export for Tableau
        const tableauFile = './nodejs_metrics_tableau.csv';
        console.log('\nExporting to Tableau CSV: ' + tableauFile);
        metrics.exportTableauCsv(tableauFile);

        // Export for PowerBI
        const powerbiFile = './nodejs_metrics_powerbi.json';
        console.log('Exporting to PowerBI JSON: ' + powerbiFile);
        metrics.exportPowerBiJson(powerbiFile);

        console.log('\nExample completed successfully!');

    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

main();
